lib/copy_field_offset.hpp

```diff
@@ -0,0 +1,46 @@
+#include <kernels/copy_field_offset.cuh>
+#include <tunable_nd.h>
+
+namespace quda {
+
+  template <class Arg> class CopyFieldOffset : public TunableKernel3D
+  {
+    using Field = typename Arg::Field;
+
+  protected:
+    Arg &arg;
+    const Field &meta;
+
+    long long bytes() const
+    {
+      return 2ll * (arg.mode == QudaOffsetCopyMode::COLLECT ? arg.in.Bytes() : arg.out.Bytes());
+    }
+
+    unsigned int minThreads() const { return arg.volume_4d_cb; }
+    int blockStep() const { return 4; }
+    int blockMin() const { return 4; }
+
+  public:
+    CopyFieldOffset(Arg &arg, const Field &meta) :
+      TunableKernel3D(meta, arg.Ls, arg.nParity),
+      arg(arg),
+      meta(meta)
+    {
+      char tmp[TuneKey::aux_n];
+      strcpy(tmp, aux);
+      writeAuxString("%s,(%d,%d,%d,%d)->(%d,%d,%d,%d),Ls=%d,nParity=%d,%s,offset=%d%d%d%d", tmp,
+                     (int)arg.dim_in[0], (int)arg.dim_in[1], (int)arg.dim_in[2], (int)arg.dim_in[3],
```

Comment:
C-style casts. How about we think of pickinh up https://github.com/fmtlib/fmt ?

Reply:
This seems like a reasonable idea.  Looking at the link, it seems that it would also be especially useful for fast conversion to strings, replacing all use of `sprintf` in the auto tuners key construction.

Reply:
yes. Separate PR though. #1176

Reply:
Switch to `static_cast` cf571ef8c.